File: pathfinding algorithm/dijkstra.py
import heapq  

def algorithm(maze, start, end):
    neighbours = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    vistied = set()
    came_from = {}
    
    # keeps track of the shortest distance between the start node to other nodes
    gscore = {start: 0}
    
    # using a heap queue
    oheap = []
    heapq.heappush(oheap, (gscore[start], start))
    
    # loop until queue is empty
    while oheap:
        current = heapq.heappop(oheap)[1]
        
        # we have reached the end node
        if current == end:
            data = []
            # reconstruct the path using came_from
            while current in came_from:
                data.append(current)
                current = came_from[current]
            data.append(start)
            # Return the path in the correct order (from start to end)
            return data[::-1]
        
        vistied.add(current)
        
        # Check all possible movements
        for i, j in neighbours:
            neighbour = current[0] + i, current[1] + j
            tentative_g_score = gscore[current] + 1  
            
            # check if the neighbouring node is a possible path
            if 0 <= neighbour[0] < len(maze[0]): 
                if 0 <= neighbour[1] < len(maze):  
                    if maze[neighbour[1]][neighbour[0]] == 1:  
                        continue
                else:
                    continue
            else:
                continue
                
            # if we have visited the neighbour but the path isnt shorter, we skip it
            if neighbour in vistied and tentative_g_score >= gscore.get(neighbour, 0):
                continue
                
            # If the new path to the neighbor is shorter or the neighbor is not in the heap
            if tentative_g_score < gscore.get(neighbour, 0) or neighbour not in [i[1] for i in oheap]:
                came_from[neighbour] = current  
                gscore[neighbour] = tentative_g_score 
                # Add the neighbor to the heap with its g-score as the priority
                heapq.heappush(oheap, (gscore[neighbour], neighbour))
                
    # if no path is found
    return False

File: pathfinding algorithm/Dijkstra.c
#include "dijkstra.h"
#include <stdio.h>
#include <limits.h>
#include <stdlib.h>

#define ROWS 15  
#define COLS 19  

// Pointer to represent x and y position in maze
typedef struct {
    int x, y;
} Point;

// Struct to represent each node in the maze
typedef struct {
    int distance;    
    Point prev;     
    bool visited;    
} Node;

// initiallise all nodes
void initialize_nodes(Node nodes[ROWS][COLS], unsigned char map[ROWS][COLS]) {
    for (int y = 0; y < ROWS; y++) {  
        for (int x = 0; x < COLS; x++) {
            nodes[y][x].distance = INT_MAX;
            nodes[y][x].prev.x = -1; 
            nodes[y][x].prev.y = -1; 
            nodes[y][x].visited = false; 
        }
    }
}

// just checking if the next node is a valid place to move
bool is_valid(int x, int y, unsigned char map[ROWS][COLS]) {
    return (x >= 0 && x < COLS && y >= 0 && y < ROWS && map[y][x] != 1);  
    }


void dijkstra(unsigned char map[ROWS][COLS], unsigned char start_x, unsigned char start_y, unsigned char dest_x, unsigned char dest_y, unsigned char path[ROWS][COLS], int *shortest_length) {
    Node nodes[ROWS][COLS];  
    initialize_nodes(nodes, map);  

    nodes[start_y][start_x].distance = 0;  

    // gives our possible directions
    Point directions[4] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    
    Point queue[ROWS * COLS];  
    int queue_start = 0, queue_end = 0;  
    
    // adds the start node to the queue
    queue[queue_end++] = (Point){start_x, start_y};  

    // loops until the queue is empty
    while (queue_start < queue_end) {
        Point current = queue[queue_start++];  
        int cx = current.x;
        int cy = current.y;  

        // if the start node and end node match, we have found the shortest path
        if (cx == dest_x && cy == dest_y) {
            break;
        }

        // skip node if its been visited
        if (nodes[cy][cx].visited) {
            continue;
        }

        // makes the current node as visited
        nodes[cy][cx].visited = true;  

        // going through all possible directions
        for (int i = 0; i < 4; i++) {
            int nx = cx + directions[i].x;  
            int ny = cy + directions[i].y;  

            // Check if the new position is valid and hasn't been visited
            if (is_valid(nx, ny, map) && !nodes[ny][nx].visited) {
                int new_dist = nodes[cy][cx].distance + 1;  

                // check if the new distance is shorter and update the distance and add to queue
                if (new_dist < nodes[ny][nx].distance) {
                    nodes[ny][nx].distance = new_dist;  
                    nodes[ny][nx].prev.x = cx;  
                    nodes[ny][nx].prev.y = cy;  
                    queue[queue_end++] = (Point){nx, ny};  
                }
            }
        }
    }

    int path_length = nodes[dest_y][dest_x].distance;  
    
    // if no possible path was found
    if (path_length == INT_MAX) {
        printf("No path found.\n");  
        path_length = -1;  
    } else {
        
        // copy the maze into the path  
        for (int y = 0; y < ROWS; y++) {
            for (int x = 0; x < COLS; x++) {
                path[y][x] = map[y][x];
            }
        }

        // update the path showing the shortest possible distance to the end node
        Point current = (Point){dest_x, dest_y};
        while (current.x != start_x || current.y != start_y) {
            path[current.y][current.x] = 9;  
            current = nodes[current.y][current.x].prev; 
        }
        path[start_y][start_x] = 2; 
    }

    *shortest_length = path_length;  
}
